<a href="https://colab.research.google.com/github/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/blob/main/Python%20NoteBooks/CASO_REGRESION_LINEAL_SIMPLE_DATOS_MINUTOS_EJERCICO_Y_PRESION_ARTERIAL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objetivo

Implementar y evaluar modelos de regresión lineal simple (RLS).


# Descripción



## Cargar librerías

In [13]:
import pandas as pd # Para procesar archuvos csv
import numpy as np # Para estructuiras de datos
import matplotlib.pyplot as plt # Para visualizar datos
import seaborn as sns # Para visuazliar datos alternativa
from sklearn.model_selection import train_test_split # Partir datos
from sklearn.linear_model import LinearRegression # PAra construir modelo RLS
from sklearn.metrics import mean_squared_error, r2_score # Métricas de modelos

## Cargar funciones

In [14]:
def f_cargar_datos(ruta_archivo):
    #------------------------------------------------------------
    # Importar datos desde un archivo CSV.
    # Argumentos:
    #   ruta_archivo: ruta del archivo a cargar.
    # Retorna:
    #   DataFrame listo para análisis.
    #------------------------------------------------------------

    datos = pd.read_csv(ruta_archivo)
    return datos

def f_visualizar_head_tail_reducido_word(datos, n=10):
    #------------------------------------------------------------
    # Mostrar primeros n y últimos n registros en una sola tabla
    # con columnas reducidas:
    #   - primeras 4 columnas
    #   - últimas 3 columnas
    #------------------------------------------------------------

    total_columnas = datos.shape[1]

    idx_prim = list(range(min(4, total_columnas)))
    idx_ult = list(range(max(total_columnas - 3, 0), total_columnas))

    columnas = sorted(set(idx_prim + idx_ult))

    datos_reducidos = datos.iloc[:, columnas]

    head_datos = datos_reducidos.head(n).astype(str)
    tail_datos = datos_reducidos.tail(n).astype(str)

    fila_puntos = pd.DataFrame([["..."] * datos_reducidos.shape[1]],
                              columns=datos_reducidos.columns)

    tabla_final = pd.concat([head_datos, fila_puntos, tail_datos])

    return tabla_final

def f_describir_datos(datos):
    #------------------------------------------------------------
    # Generar estadísticas descriptivas básicas
    #------------------------------------------------------------

    describe = datos.describe()

    structure = datos.info()

    return {
        "describe": describe,
        "structure": structure
    }

def f_particionar_datos(datos, proporcion_entrenamiento=0.7):
    #------------------------------------------------------------
    # Dividir dataset en entrenamiento y validación
    #------------------------------------------------------------

    train, test = train_test_split(
        datos,
        train_size=proporcion_entrenamiento,
        random_state=2026
    )

    return {
        "datos_entrenamiento": train,
        "datos_validacion": test
    }

from sklearn.linear_model import LinearRegression

def f_construir_modelo(datos_entrenamiento, variable_independiente, variable_dependiente):
    #------------------------------------------------------------
    # f_construir_modelo()
    #
    # Objetivo:
    #   Construir un modelo de regresión lineal simple
    #
    # Descripción:
    #   La función toma un conjunto de datos de entrenamiento
    #   y ajusta un modelo de regresión lineal simple usando
    #   una variable independiente y una dependiente.
    #
    # Argumentos:
    #   datos_entrenamiento      : DataFrame de entrenamiento
    #   variable_independiente   : nombre de la variable X
    #   variable_dependiente     : nombre de la variable Y
    #
    # Retorna:
    #   modelo entrenado listo para predicción y evaluación
    #------------------------------------------------------------

    # 1. Definir variables
    X = datos_entrenamiento[[variable_independiente]]
    y = datos_entrenamiento[variable_dependiente]

    # 2. Crear modelo
    modelo = LinearRegression()

    # 3. Entrenar modelo
    modelo.fit(X, y)

    # 4. Retornar modelo
    return modelo

def f_diagrama_dispersion_tendencia(modelo, datos, x, y):
    #------------------------------------------------------------
    # Visualiza dispersión + línea de tendencia
    # Incluye r y R² en subtítulo
    #------------------------------------------------------------

    x_vals = datos[x]
    y_vals = datos[y]

    # correlación
    r = np.corrcoef(x_vals, y_vals)[0,1]

    # predicción modelo
    y_pred = modelo.predict(datos[[x]])

    # R²
    r2 = r2_score(y_vals, y_pred)

    plt.figure(figsize=(8,6))

    plt.scatter(x_vals, y_vals)

    # ordenar para línea
    x_sorted = np.sort(x_vals)
    y_line = modelo.predict(x_sorted.reshape(-1,1))

    plt.plot(x_sorted, y_line)

    plt.title("Dispersión y tendencia")
    plt.suptitle(f"{x} vs {y} ; r={r:.3f} ; R²={r2:.3f}")

    plt.xlabel(x)
    plt.ylabel(y)

    plt.show()

def f_evaluar_modelo(modelo, datos_validacion, variable_dependiente, variable_independiente):
    #------------------------------------------------------------
    # Evaluar modelo usando datos de validación
    #------------------------------------------------------------

    X = datos_validacion[[variable_independiente]]
    y_real = datos_validacion[variable_dependiente]

    pred = modelo.predict(X)

    mse = mean_squared_error(y_real, pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_real, pred)

    # R² ajustado
    n = len(y_real)
    p = 1


    r2_adj = 1 - (1 - r2)*(n - 1)/(n - p - 1)

    resultado = pd.DataFrame({
        "R_square": [round(r2,4)],
        "R_square_ajustado": [round(r2_adj,4)],
        "MSE": [round(mse,4)],
        "RMSE": [round(rmse,4)]
    })

    return resultado

Importar funciones directamente de cada enlace de cada archivo de datos.


In [29]:
import requests

url = "https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/Python%20NoteBooks/funciones/funciones_RLS.py"

exec(requests.get(url).text)

## Cargar datos

Se ejecuta la función *f_cargar_datos()* indicando la ruta del archivo.


In [30]:
datos_1 = f_cargar_datos("https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/datos/datos_ejercicio_presion_ruido_3000.csv")
datos_2 = f_cargar_datos("https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/datos/datos_ejercicio_presion_didactico_3000.csv")

## Visualizar datos_1

Los primeros y últimos datos de *datos_1*

In [31]:
f_visualizar_head_tail_reducido_word(datos_1)

,actividad_min,presion_sistolica
0,418,138
1,172,142
2,136,150
3,331,139
4,432,139
5,254,140
6,588,135
7,411,140
8,289,142
9,235,135


## Visualizar datos_2

Los primeros y últimos datos de *datos_2*

In [32]:
f_visualizar_head_tail_reducido_word(datos_2)

,actividad_min,presion_sistolica
0,418,135
1,172,142
2,136,146
3,331,140
4,432,133
5,254,146
6,588,136
7,411,138
8,289,140
9,235,146


## Describir datos_1

In [33]:
f_describir_datos(datos_1)

{'describe':        actividad_min  presion_sistolica
 count    3000.000000        3000.000000
 mean      298.430333         141.224000
 std       174.632756           6.252428
 min         0.000000         122.000000
 25%       145.000000         137.000000
 50%       299.000000         141.000000
 75%       448.000000         145.000000
 max       600.000000         164.000000,
 'structure': "<class 'pandas.core.frame.DataFrame'>\nRangeIndex: 3000 entries, 0 to 2999\nData columns (total 2 columns):\n #   Column             Non-Null Count  Dtype\n---  ------             --------------  -----\n 0   actividad_min      3000 non-null   int64\n 1   presion_sistolica  3000 non-null   int64\ndtypes: int64(2)\nmemory usage: 47.0 KB\n"}

## Describir datos_2

In [34]:
f_describir_datos(datos_2)

{'describe':        actividad_min  presion_sistolica
 count    3000.000000        3000.000000
 mean      298.430333         141.036000
 std       174.632756           4.862501
 min         0.000000         127.000000
 25%       145.000000         137.000000
 50%       299.000000         140.000000
 75%       448.000000         144.000000
 max       600.000000         158.000000,
 'structure': "<class 'pandas.core.frame.DataFrame'>\nRangeIndex: 3000 entries, 0 to 2999\nData columns (total 2 columns):\n #   Column             Non-Null Count  Dtype\n---  ------             --------------  -----\n 0   actividad_min      3000 non-null   int64\n 1   presion_sistolica  3000 non-null   int64\ndtypes: int64(2)\nmemory usage: 47.0 KB\n"}

Para ambos conjuntos de datos el número de observaciones es el mismo, los estadísticos de tendencia central y dispérsi´pon son simialres lo que varía son los valores de la variables inbdependiente y dependiente.

Ahora se implementa y se evalúa el modelo de regresión en el desarrollo, dos modelos para ambos conjuntos de datos.

# Desarrollo

Se manda llamar la implementación del modelo con los dos conjuntos de datos


## Modelo de regresión lineal simple (RLS)

### Particonar datos_1 para MRL con datos_1


In [35]:
# Particionar datos_1
particion = f_particionar_datos(datos_1)

datos_entrenamiento_1 = particion["datos_entrenamiento"]
datos_validacion_1 = particion["datos_validacion"]

f_visualizar_head_tail_reducido_word(datos_entrenamiento_1)


,actividad_min,presion_sistolica
1876,486,133
2240,81,137
1118,94,149
2898,26,154
1595,400,139
903,117,144
1109,454,143
1083,47,142
2006,399,143
1576,335,139


In [22]:
f_visualizar_head_tail_reducido_word(datos_validacion_1)

,actividad_min,presion_sistolica
146,445,138
935,203,149
11,437,134
2450,340,139
1055,27,141
2476,326,137
2688,458,145
754,195,137
2177,205,145
384,413,135


### Construior MRL para datos_1

In [37]:
# Construir modelo
modelo1_RLS = f_construir_modelo(
    datos_entrenamiento_1,
    "actividad_min",
    "presion_sistolica"
)

### Evaluar MRL para datos_1

In [38]:
resultado_1 = f_evaluar_modelo(
    modelo1_RLS,
    datos_validacion_1,
    "presion_sistolica",
    "actividad_min"
)

resultado_1

,R_square,R_square_ajustado,MSE,RMSE
0,0.3236,0.3228,27.0077,5.1969


### Particionar datos_2 para MRL con datos_2

In [41]:
# Particionar datos_2
particion = f_particionar_datos(datos_2)

datos_entrenamiento_2 = particion["datos_entrenamiento"]
datos_validacion_2 = particion["datos_validacion"]

f_visualizar_head_tail_reducido_word(datos_entrenamiento_2)

,actividad_min,presion_sistolica
1876,486,133
2240,81,146
1118,94,143
2898,26,150
1595,400,140
903,117,144
1109,454,140
1083,47,148
2006,399,136
1576,335,137


In [40]:
f_visualizar_head_tail_reducido_word(datos_validacion_2)

,actividad_min,presion_sistolica
146,445,136
935,203,144
11,437,135
2450,340,140
1055,27,153
2476,326,140
2688,458,136
754,195,142
2177,205,144
384,413,137


### Construior MRL para datos_2

In [45]:
# Construir modelo
modelo2_RLS = f_construir_modelo(
    datos_entrenamiento_2,
    "actividad_min",
    "presion_sistolica"
)

### Evaluar MRL para datos_2

In [46]:
resultado_2 = f_evaluar_modelo(
    modelo2_RLS,
    datos_validacion_2,
    "presion_sistolica",
    "actividad_min"
)

resultado_2

,R_square,R_square_ajustado,MSE,RMSE
0,0.5474,0.5469,10.8026,3.2867


# Interpretación

Se analizaron dos conjuntos de datos arrojando de manera similar estadíscitos de tedencia central y de dispersión.

El caso para su revisón se encuentra en el enlace en la plataforma google collab siguiente: https://colab.research.google.com/drive/1A9wv-u7ZDgepJiGO53V3i9e5TRMlsVZV?usp=sharing y también se puede localizar en el enlace de la platgaforma *github* del autor de uso personal siguiente:

El caso utiliza funciones previamente programadas que se encuentran para su descarga y ejecución en el enlace siguiente: https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/Python%20NoteBooks/funciones/funciones_RLS.py

La evalaución del primer modelo con los *datos_1* que tienen ruido se identifica lo siguiente:

* El valor de *r square* fué de *0.3236* que significa que en ese modelo y con esos datos, la variable independiente caminata en minutos semanales, representa aproximadamente el 32% de la presión sistólica de una persona.

* Luego el valor de *RMSE* que representa que tantas unidades varían las predicciones con respecto a los valores reales fue de aproximadamente *5.19*.

La evalaución del segundo modelo con los *datos_2* que sirven para comparación y descripc´n académica, se identifica lo siguiente:

* El valor de *r square* fué de *0.5474* que significa que en ese modelo y con esos datos, la variable independiente caminata en minutos semanales, representa aproximadamente el 54% de la presión sistólica de una persona que resulta en un modelo de mejor calidad con esto datos.

* Luego el valor de *RMSE* que representa que tantas unidades varían las predicciones con respecto a los valores reales fue de aproximadamente *3.28*, lo que lo hace más eficiente.

Estos datos que arrojan tanto los estadísticos de evaluación r square como *RMSE* son similares los que arroja tanto *R* como *Python*, lo que finaliza con esto, el caso de implementación y evaluación de regresión lineal simple para un mismo conjunto de datos.

